extensions/s3/src/main/java/io/deephaven/extensions/s3/S3OutputStream.java

```diff
@@ -0,0 +1,270 @@
+         * The part number for the part to be uploaded
+         */
+        private int partNumber;
+
+        /**
+         * The future for the part upload
+         */
+        private CompletableFuture<UploadPartResponse> future;
+
+        OutgoingRequest(final int partSize) {
+            buffer = ByteBuffer.allocate(partSize);
+            partNumber = INVALID_PART_NUMBER;
+        }
+    }
+
+    private String initiateMultipartUpload() throws IOException {
+        final CreateMultipartUploadRequest createMultipartUploadRequest = CreateMultipartUploadRequest.builder()
+                .bucket(uri.bucket().orElseThrow())
+                .key(uri.key().orElseThrow())
+                .build();
+        final CompletableFuture<CreateMultipartUploadResponse> future =
+                s3AsyncClient.createMultipartUpload(createMultipartUploadRequest);
+        final CreateMultipartUploadResponse response;
+        try {
+            response = future.get();
+        } catch (final InterruptedException | ExecutionException | CancellationException e) {
+            throw handleS3Exception(e, String.format("initiating multipart upload for uri %s", uri), s3Instructions);
+        }
+        return response.uploadId();
+    }
+
+    /**
+     * Send a part request for the given buffer. This method assumes that the buffer is non-empty.
+     */
+    private void sendPartRequest(final OutgoingRequest request) throws IOException {
+        if (nextPartNumber > MAX_PART_NUMBER) {
+            throw new IOException("Cannot upload more than " + MAX_PART_NUMBER + " parts for uri " + uri + ", please" +
+                    " try again with a larger part size");
+        }
+        if (request.future != null) {
+            throw new IllegalStateException("Request already in progress for uri " + uri + " with part number " +
+                    nextPartNumber);
+        }
+        final UploadPartRequest uploadPartRequest = UploadPartRequest.builder()
+                .bucket(uri.bucket().orElseThrow())
+                .key(uri.key().orElseThrow())
+                .uploadId(uploadId)
+                .partNumber(nextPartNumber)
+                .build();
+        request.buffer.flip();
+        request.future = s3AsyncClient.uploadPart(uploadPartRequest,
+                AsyncRequestBody.fromByteBufferUnsafe(request.buffer));
+        request.partNumber = nextPartNumber;
+        nextPartNumber++;
+    }
+
+    private void waitForCompletion(final OutgoingRequest request) throws IOException {
+        final UploadPartResponse uploadPartResponse;
+        try {
+            uploadPartResponse = request.future.get();
+        } catch (final InterruptedException | ExecutionException | CancellationException e) {
+            throw handleS3Exception(e, String.format("waiting for part %d for uri %s to complete uploading",
+                    request.partNumber, uri), s3Instructions);
+        }
+
+        completedParts.add(CompletedPart.builder()
```

Comment:
Looks like `software.amazon.awssdk.services.s3.internal.multipart.SdkPojoConversionUtils#toCompletedPart(software.amazon.awssdk.services.s3.model.UploadPartResponse, int)` might be the more correct way to construct CompletedPart?